dali/pipeline/data/buffer_manager.h

```diff
@@ -0,0 +1,179 @@
+    const size_t buffer_initial_size = 1048576;
+
+    cpu_buffers_.resize(buffers_init_size);
+    pinned_buffers_.resize(buffers_init_size);
+    gpu_buffers_.resize(buffers_init_size);
+
+    for (int i = 0; i < buffers_init_size; ++i) {
+      cpu_buffers_[i].reset(new Buffer<CPUBackend>);
+      cpu_buffers_[i]->Resize({buffer_initial_size});
+
+      pinned_buffers_[i].reset(new Buffer<CPUBackend>);
+      pinned_buffers_[i]->set_pinned(true);
+      pinned_buffers_[i]->Resize({buffer_initial_size});
+
+      gpu_buffers_[i].reset(new Buffer<GPUBackend>);
+      gpu_buffers_[i]->Resize({buffer_initial_size});
+    }
+
+    CUDA_CALL(cudaSetDevice(original_device));
+  }
+
+  // Acquire a buffer
+  inline unique_ptr<Buffer<CPUBackend>> AcquireBuffer(const size_t nbytes,
+                                                      const bool pinned) override;
+  inline unique_ptr<Buffer<GPUBackend>> AcquireBuffer(const size_t nbytes) override;
+
+  // Release a buffer
+  inline void ReleaseBuffer(unique_ptr<Buffer<CPUBackend>> *buffer,
+                            const bool pinned) override;
+  inline void ReleaseBuffer(unique_ptr<Buffer<GPUBackend>> *buffer) override;
+
+ private:
+  // Buffer lists for re-use
+  vector<unique_ptr<Buffer<CPUBackend>>> cpu_buffers_;
+  vector<unique_ptr<Buffer<GPUBackend>>> gpu_buffers_;
+  vector<unique_ptr<Buffer<CPUBackend>>> pinned_buffers_;
+};
+
+unique_ptr<Buffer<CPUBackend>> LinearBufferManager::AcquireBuffer(const size_t bytes,
+                                                                  const bool pinned) {
+  if (pinned) {
+    std::lock_guard<std::mutex> lock(pinned_buffers_mutex_);
+    return LinearSearch(&pinned_buffers_, bytes, true);
+  } else {
+    std::lock_guard<std::mutex> lock(cpu_buffers_mutex_);
+    return LinearSearch(&cpu_buffers_, bytes);
+  }
+}
+
+void LinearBufferManager::ReleaseBuffer(unique_ptr<Buffer<CPUBackend>> *buffer, bool pinned) {
+  if (pinned) {
+    std::lock_guard<std::mutex> lock(pinned_buffers_mutex_);
+    pinned_buffers_.push_back(std::move(*buffer));
+  } else {
+    std::lock_guard<std::mutex> lock(cpu_buffers_mutex_);
+    cpu_buffers_.push_back(std::move(*buffer));
+  }
+}
+
+unique_ptr<Buffer<GPUBackend>> LinearBufferManager::AcquireBuffer(const size_t bytes) {
+  std::lock_guard<std::mutex> lock(gpu_buffers_mutex_);
+  std::cout << "Acquiring buffer of size " << bytes << std::endl;
+  std::cout << "Size of buffer queue: " << gpu_buffers_.size() << std::endl;
+  auto buf = LinearSearch(&gpu_buffers_, bytes);
+  std::cout << "I will give buffer " << buf->data_.get() << std::endl;
+  if (buf->data_.get()) {
+    std::cout << "Buffer " << buf->data_.get() << " has " << buf->capacity() << " bytes." << std::endl;
+  }
+  return buf;
+}
+
+void LinearBufferManager::ReleaseBuffer(unique_ptr<Buffer<GPUBackend>> *buffer) {
+  std::lock_guard<std::mutex> lock(gpu_buffers_mutex_);
+  std::cout << "Releasing buffer " << buffer->get()->data_.get() << std::endl;
```

Comment:
How about wrapping that into some #ifdef BUFFER_MAMANGER_DEBUG macro?

Reply:
As written in OP in things left to do: remove all debug couts and DALI_WARN